### ISSUE TITLE
ToCloudEventInternalAsync fails when using ContentMode.Binary in SystemTextJson and NewtonsoftJson

### ISSUE THREAD

**Issue by squanbeck** [cloudevents/sdk-csharp#221]
Following this sample of code from the guide,

https://github.com/cloudevents/sdk-csharp/blob/3ce4aa08f2763cfd68d5e1a0865ac9442e5b8021/docs/guide.md?plain=1#L293-L303

https://github.com/cloudevents/sdk-csharp/blob/3ce4aa08f2763cfd68d5e1a0865ac9442e5b8021/docs/guide.md?plain=1#L314-L334

https://github.com/cloudevents/sdk-csharp/blob/3ce4aa08f2763cfd68d5e1a0865ac9442e5b8021/docs/guide.md?plain=1#L366-L368

I end up with this error (line 151): 

https://github.com/cloudevents/sdk-csharp/blob/3ce4aa08f2763cfd68d5e1a0865ac9442e5b8021/src/CloudNative.CloudEvents/Http/HttpClientExtensions.cs#L139-L152

When I looked further into it, it seems that the variable, "headers", is empty when using ContentMode.Binary. This is the case when using both SystemTextJson and NewtonsoftJson (likely when using the other packages as well, but these are the only ones I tested). However, I did notice that "content.headers" does contain the ce-specversion. Now, I am unsure if I am reading the CloudEvents specification correctly, but I think that this is in conformance, based off of my understanding that in binary mode, the event data need not be in the HTTP headers themselves, but rather in the message (in this case, the "content") metadata. 

https://github.com/cloudevents/spec/blob/604fef3eea26801d1249bd7ee796e947d38cb5a1/cloudevents/spec.md?plain=1#L125-L134

If my understanding is accurate, does an extra check for "content.headers" containing ce-specversion need to exist here, somewhere around line 147?

**Comment by jskeet** [comment]
I'll have a look at this in detail next week - although I'm travelling, which may delay things.

**Comment by squanbeck** [comment]
Sounds good, thank you!

**Comment by jskeet** [comment]
(Looking now briefly as I have some time.)

What's missing from this issue is how you're receiving the request, and ideally a repro of the problem. I've never been 100% clear how .NET decides which headers in an inbound request should be counted as content headers vs which are just request headers. Are you encountering the problem when receiving an *actual* HTTP request on a server, or is this for an `HttpMessage` you've constructed in code?

It's entirely possible that we should indeed look at all headers together - but I'll need to look in more detail, and if you can provide more context of where you're seeing this, it would help.

**Comment by jskeet** [comment]
Note to self: RFC-2616 never talks about content headers, but it does include entity headers, which appear to include "everything not explicitly declared as a general header or request header".

**Comment by squanbeck** [comment]
As you have guessed, this is _just_ the contrived example from the guide; it does not involve any _actual_ HTTP request on a server. To be perfectly honest, I don't know that I understand the HTTP protocol well enough to know how the request headers and content/entity headers differ, particularly in their relation to .NET's implementation.

**Comment by jskeet** [comment]
Ah, I see what you mean - I hadn't seen why you'd included all the guide stuff before. Right, that makes sense. I'll investigate further next week and see whether we need a code change, a guide change, or both...

**Comment by jskeet** [comment]
Will look into that when I look at the whole thing, hopefully early next week :)

**Comment by squanbeck** [comment]
So I came back to this for a little bit of a further look today. Seems I only ran into this issue as I misunderstood how the HttpRequestMessage and HttpRequest were being used. When using System.Web.Http (I think this only works with .NET Framework), and using a controller inheriting from ApiController, the HttpRequestMessage is available from the inherited member "Request". When using the newer(?) Microsoft.AspNetCore.Mvc and controllers inheriting from ControllerBase, the HttpRequest is available from the inherited member "Request". This indicates that I need to use the CloudNative.CloudEvents.AspNetCore library to access the HttpRequest.ToCloudEventAsync() method. I got confused within the guide, as I had assumed that the "request" in line 367 was an HttpRequestMessage instead of an HttpRequest and I was using .NET Core to test this all out. Hope that helps!
